[proposals] Threads

andreaTP opened the pull request:
Opening for visibility, just implemented a good chunk of validation, the opcodes are still missing.

Notes:
- working in the interpreter now
- runs `jq` and `protoc` from `wasilibs`
- no implementation of `atomic.fence` as it's completely untested
- the testsuite is really poor

andreaTP commented on wasi-tests/src/test/java/com/dylibso/chicory/wasi/WasiPreview1Test.java:
```diff
@@ -443,4 +446,53 @@ private static WasiPreview1 wasiWithDirectory(String guest, Path host) {
+                var module =
+                        Parser.parse(
+                                new File(
+                                        "/home/andreatp/workspace/go-protoc-gen-grpc-java/internal/wasm/protoc-gen-grpc-java.wasm"));
```
My goal was to run this module, coming from:
https://github.com/wasilibs/go-protoc-gen-grpc-java

But something is not working yet :cry:

andreaTP replied:
Working now! Both protoc and jq

andreaTP commented:
This is ready for review, is the full interpreter implementation.
I'm already working on porting to the compiler :slightly_smiling_face:

chirino commented:
given atomic fence is not implemented.  What happens right now if a wasm module uses that instruction?  I hope we give the user a decent error message that the wasm module is using a unsupported instruction?

chirino commented on runtime/src/main/java/com/dylibso/chicory/runtime/ByteBufferMemory.java:
```diff
@@ -44,6 +47,133 @@ public ByteBufferMemory(MemoryLimits limits, MemAllocStrategy allocStrategy) {
+                monitors.get(address).wait(millis, nanos);
+            }
+        } catch (InterruptedException ie) {
+            throw new ChicoryInterruptedException("Thread interrupted");
```
Maybe added `Thread.currentThread().interrupt(); // Restore interrupt status`

andreaTP commented:
> given atomic fence is not implemented. What happens right now if a wasm module uses that instruction? I hope we give the user a decent error message that the wasm module is using a unsupported instruction?

[This](https://github.com/dylibso/chicory/blob/cd67b24a402c607892752cb3f587bbae0f3c7566/wasm/src/main/java/com/dylibso/chicory/wasm/Validator.java#L1889-L1890) will be the first error that shows up to a user, we can improve the text if you think that is not enough(something like: "open an issue at ...").
So far I haven't seen any usage of it, and the spec is a bit to vague to try to anticipate a possible implementation without examples and tests ... I left it out on purpose.

andreaTP commented on runtime/src/main/java/com/dylibso/chicory/runtime/InterpreterMachine.java:
```diff
@@ -1995,6 +2165,321 @@ private static void LOCAL_TEE(MStack stack, Operands operands, StackFrame curren
+
+    private static void I64_ATOMIC_STORE8(MStack stack, Instance instance, Operands operands) {
+        var value = (byte) stack.pop();
+        var ptr = (int) (operands.get(1) + (int) stack.pop());
```
good catch :+1:

andreaTP commented on runtime/src/main/java/com/dylibso/chicory/runtime/ByteArrayMemory.java:
```diff
@@ -58,6 +61,133 @@ public ByteArrayMemory(MemoryLimits limits, MemAllocStrategy allocStrategy) {
+        }
+        if (monitors.containsKey(address) && monitors.get(address).decrementAndGet() == 0) {
+            monitors.remove(address);
+        }
```
I'll change it :+1:

andreaTP commented on runtime/src/main/java/com/dylibso/chicory/runtime/ByteArrayMemory.java:
```diff
@@ -58,6 +61,133 @@ public ByteArrayMemory(MemoryLimits limits, MemAllocStrategy allocStrategy) {
+                monitors.get(address).wait(millis, nanos);
+            }
+        } catch (InterruptedException ie) {
+            throw new ChicoryInterruptedException("Thread interrupted");
```
Make sense :+1:

andreaTP commented:
I'll go through the review comments now, I added the implementation in the compiler as a single commit, is just a lot of very repetitive code, but I can take it out for a follow up PR when requested.

andreaTP commented on runtime/src/main/java/com/dylibso/chicory/runtime/ByteArrayMemory.java:
```diff
@@ -58,6 +61,133 @@ public ByteArrayMemory(MemoryLimits limits, MemAllocStrategy allocStrategy) {
+    // Notify all waiters at this address
+    @Override
+    public int notify(int address, int maxThreads) {
+        if (!shared()) {
+            return 0;
+        }
```
I reviewed the spec and the code, the "atomic" opcodes acting on a non-shared memory should, usually, behave as the corresponding non atomic operation.
`wait` instead should trap, I've added the relevant code, but I don't have a great feeling as there are no tests nor examples that can be used to drive the implementation.

I think this is enough as a starting point, and we can iterate when we receive a bug/issue.

andreaTP commented:
Ready for the next review round!

- addressed all of the comments (thanks @chirino :pray: )
- added the full implementation in the compiler
- added all the relevant tests

chirino commented on compiler/src/main/java/com/dylibso/chicory/compiler/internal/Compiler.java:
```diff
@@ -561,7 +562,7 @@ private static void emitFunction(
 
         // uncomment if you ever want to troubleshoot invalid bytecode generation
-        // methodWriter = new CheckMethodAdapter(methodWriter);
+        methodWriter = new CheckMethodAdapter(methodWriter);
 
```
I assume you uncommented this to aid development.

andreaTP replied:
Yes, that's a day saver, thanks for it!
Good catch!

chirino reviewed (approved):
just one small nit but...  LGTM!
